### ISSUE TITLE
call to Expression->render() is lacking somewhere

### ISSUE THREAD

**Issue by PhilippGrashoff** [atk4/dsql#155]
Hi,
I am sorry to not being able to describe this issue perfectly. After many hours of searching, I found that a call to Expression->render() seems to be missing somewhere. As the explanation is very complex, I made a screencast: [not available any more]

Here is the SQL expressions that failed if render was not called:
```
        $this->addExpression('sum_amount_guests', [
            $this->expr("(SELECT SUM(`group`.`amount_guests_payfull` + `group`.`amount_guests_payreduced`) FROM `group` WHERE `group`.`id` IN (SELECT `group_to_tour`.`group_id` FROM `group_to_tour` WHERE `group_to_tour`.`tour_id` = [id]) AND `group`.`status` IN ('booked' , 'completed'))"),
            'system' => true,
            'never_save' => true,
        ]);
        $this->addExpression('sum_option_guests', [
            $this->expr("(SELECT SUM(`amount_guests_payfull` + `amount_guests_payreduced`) FROM `group` WHERE `group`.`id` IN (SELECT `group_id` FROM `group_to_tour` WHERE `group_to_tour`.`tour_id` = [id]) AND `group`.`status` IN ('option'))"),
            'system' => true,
            'never_save' => true,
        ]);
```

**Comment by PhilippGrashoff** [comment]
I think I found the place where its missing: In Expression.php Line 620.
In the if-part `if ($connection instanceof \PDO) {` `$this->render()` is called, while in the else part it is not. Adding `$this->render()` to the else part as well did work for me.

**Comment by DarkSide666** [comment]
Well I think it's fine as it is now.

If it's using PDO connection then it renders Expression and do some stuff.
If it's not using PDO connection but one of \atk4\dsql\Connection classes then that connection class is responsible of calling expression render.

Can you show me what Connection do you use? How it is configured?

**Comment by PhilippGrashoff** [comment]
Hi, thank you for the reply, will test if that works. One thing about this issue is that it was very very difficult to debug, as using dumper: calls render(), so here the correct query was output instead of what really is sent. In the end it was a lucky coincidence I really found what's happening. I should add some info to docs that when using expression fields should be referenced as you showed above.

**Comment by DarkSide666** [comment]
In simple expression it would work just fine without table alias,but in this case when having nested subselects in expression definitely should use expr() arguments.
Also you could think of is it good to have such long subselect as expression, but ... why not.

**Comment by PhilippGrashoff** [comment]
Hi,

I just tested the solutions you gave with the newest releases, and they didnt work. I tested:
1) if you explicitly define $table_alias property of that model;
2) or pass field object as parameter in expr() method like this $this->expr('SELECT foo FROM table WHERE table.id=[my_id]', ['my_id' => $this->getElement('id')]

and both didnt have any effect.

However, a call to $this->render() in line 620 of Expression.php does the job :)
Maybe this is something that was undiscovered yet because no one used that complex expression?

I will read into Expression->render(), but would be happy for any working solution, maybe mapping that query into atk data?

Best regards

**Comment by PhilippGrashoff** [comment]
Hi, after stumbling over this again, I wanted to add that I had a call to expr() inside addExpression(). Without this, it works fine, like:
```
        $this->addExpression(
            'sum_amount_guests',
            [
                "(SELECT SUM(
                    IFNULL(`amount_guests_payfull`, 0) +
                    IFNULL(`amount_guests_payreduced`, 0)
                )
                FROM `group_to_tour`
                WHERE `group_to_tour`.`tour_id` = [id]
                AND `group_to_tour`.`status` IN ('booked', 'completed'))",
                'type' => 'integer'
            ]
        );
```
I still think that there is an error, but it only seems to occur with expr() inside addExpression(). So I have my workaround, just in case anybody else stumbles upon this.

**Comment by DarkSide666** [comment]
Is it possible for you to write test case which fail so I can see exactly what's going on there and how it fails?
Is this last expression you posted fails?

**Comment by PhilippGrashoff** [comment]
Hi, no, the last expression works. The "nested" `$this->addExpression('expressionName', $this->expr());` is what fails. Its rendered without table alias in front, e.g. \`id\` instead of \`table\`.\`id\`. 

I could try to write a testcase, but it might prove difficult, as some functions that could be used to get 
the rendered query call `->render()`. e.g. its called in Expressions `__debugInfo()`. Thats why it initially took me hours to get to the problem, as I tried to get the rendered expression this way. Only when I had a look at the MySQL query log, I was able to find it :) I made a screencast explaining the problem but I deleted the link it seems, I could send it to you again if you like. 

Basically I am 100% sure its due to the "nesting", and ->render() not being called on the "nested" expr().
If thats valid code in general is another question. I will check if there are examples of $this->addExpression('expressionName', $this->expr()) or if it was something I made up accidently.

**Comment by PhilippGrashoff** [comment]
As line numbers are not up-to-date any more in my previous comments, what fixed it for me then was to add $this->render() to the small else part if connection isnt PDO:
```
else {
           $this->render(); 
           return $connection->execute($this);
        }
```

**Comment by PhilippGrashoff** [comment]
I searched and didnt find any other occurence of "nested" `$this->addExpression('expressionName`, $this->expr());`. So it seems I wrote some non-standard code which fails. Should have read examples better :) If you like you can close this, on the other hand, maybe have a thought about that call to `render()` in the else block?

**Comment by mvorisek** [comment]
@PhilippGrashoff can you create PR with test for this?

**Comment by mvorisek** [comment]
@PhilippGrashoff can you create PR with test for this?

**Comment by mvorisek** [comment]
@PhilippGrashoff provide test case
